File: r/tests/testthat/test_competition_canvas_1d.R
# test_competition_canvas_1d.R — Unit tests for 1D animate_competition_canvas.

# Fixture: 3 1D voters, 2 sticker competitors, 1 seat, 2 rounds.
# Competitor 0 at x=0 wins every round (closest to 2 of 3 voters).
# n_frames = 3 (Round 1, Round 2, Final); 1 seat holder per frame.

.voters_1d <- c(-0.5, 0.1, 1.8)   # 3 ideal points (flat 1D)

.trace_1d <- function() {
  competition_run(
    competitor_positions = c(0.0, 2.0),
    strategy_kinds       = c("sticker", "sticker"),
    voter_ideals         = .voters_1d,
    dist_config          = make_dist_config(n_dims = 1L),
    engine_config        = make_competition_engine_config(
      seat_count  = 1L,
      seat_rule   = "plurality_top_k",
      max_rounds  = 2L,
      step_config = make_competition_step_config(kind = "fixed", fixed_step_size = 0.5)
    )
  )
}

# ---------------------------------------------------------------------------
# Payload shape tests
# ---------------------------------------------------------------------------

test_that("1D payload has n_dims = 1", {
  skip_without_lib()
  trace <- .trace_1d()
  w <- animate_competition_canvas(trace, voters = .voters_1d, width = 900L, height = 400L)
  expect_equal(w$x$n_dims, 1L)
})

test_that("1D payload has voters_x and no voters_y", {
  skip_without_lib()
  trace <- .trace_1d()
  w <- animate_competition_canvas(trace, voters = .voters_1d, width = 900L, height = 400L)
  expect_true("voters_x" %in% names(w$x))
  expect_false("voters_y" %in% names(w$x))
  expect_length(w$x$voters_x, 3L)
})

test_that("1D payload positions have length-1 inner lists", {
  skip_without_lib()
  trace <- .trace_1d()
  w <- animate_competition_canvas(trace, voters = .voters_1d, width = 900L, height = 400L)
  for (frame in w$x$positions) {
    for (comp_pos in frame) {
      expect_length(comp_pos, 1L)
    }
  }
})

test_that("1D payload has no ylim", {
  skip_without_lib()
  trace <- .trace_1d()
  w <- animate_competition_canvas(trace, voters = .voters_1d, width = 900L, height = 400L)
  expect_false("ylim" %in% names(w$x))
})

test_that("1D payload has no winsets", {
  skip_without_lib()
  trace <- .trace_1d()
  w <- animate_competition_canvas(trace, voters = .voters_1d, width = 900L, height = 400L)
  expect_false("winsets" %in% names(w$x))
})

test_that("1D dim_names is a length-1 list", {
  skip_without_lib()
  trace <- .trace_1d()
  w <- animate_competition_canvas(
    trace, voters = .voters_1d, dim_names = "Policy Dimension",
    width = 900L, height = 400L
  )
  expect_length(w$x$dim_names, 1L)
  expect_equal(w$x$dim_names[[1L]], "Policy Dimension")
})

test_that("1D xlim covers all voter and competitor positions", {
  skip_without_lib()
  trace <- .trace_1d()
  w <- animate_competition_canvas(trace, voters = .voters_1d, width = 900L, height = 400L)
  xlim <- unlist(w$x$xlim)
  expect_lte(xlim[1L], min(.voters_1d))
  expect_gte(xlim[2L], max(.voters_1d))
})

# ---------------------------------------------------------------------------
# IC tests
# ---------------------------------------------------------------------------

test_that("compute_ic = FALSE leaves IC keys absent", {
  skip_without_lib()
  trace <- .trace_1d()
  w <- animate_competition_canvas(trace, voters = .voters_1d, width = 900L, height = 400L)
  expect_false("indifference_curves" %in% names(w$x))
  expect_false("ic_competitor_indices" %in% names(w$x))
})

test_that("compute_ic = TRUE produces correctly shaped 1D IC payload", {
  skip_without_lib()
  trace <- .trace_1d()
  w <- animate_competition_canvas(
    trace,
    voters     = .voters_1d,
    compute_ic = TRUE,
    width      = 900L,
    height     = 400L
  )
  d        <- trace$dims()
  n_frames <- d$n_rounds + 1L   # 3
  n_voters <- 3L

  expect_true("indifference_curves" %in% names(w$x))
  ic <- w$x$indifference_curves

  expect_length(ic, n_frames)
  for (frame in ic) {
    # Competitor 0 is the single seat holder every frame.
    expect_length(frame, 1L)
    voter_curves <- frame[[1L]]
    expect_length(voter_curves, n_voters)
    for (pts in voter_curves) {
      # Each IC is 0, 1, or 2 floats.
      expect_lte(length(pts), 2L)
    }
  }
})

test_that("1D IC boundaries are symmetric around the voter ideal", {
  # voter at 0.5, seat at 0.0 → dist = 0.5 → utility = -0.5 under linear loss.
  # level_set_1d(0.5, 1.0, -0.5) = [0.0, 1.0]
  skip_without_lib()
  trace <- competition_run(
    competitor_positions = 0.0,
    strategy_kinds       = "sticker",
    voter_ideals         = 0.5,
    dist_config          = make_dist_config(n_dims = 1L),
    engine_config        = make_competition_engine_config(
      seat_count  = 1L,
      seat_rule   = "plurality_top_k",
      max_rounds  = 1L,
      step_config = make_competition_step_config(kind = "fixed", fixed_step_size = 0.0)
    )
  )
  w <- animate_competition_canvas(
    trace,
    voters         = 0.5,
    compute_ic     = TRUE,
    ic_loss_config = make_loss_config(loss_type = "linear", sensitivity = 1.0),
    ic_dist_config = make_dist_config(n_dims = 1L),
    width          = 900L,
    height         = 400L
  )
  pts <- w$x$indifference_curves[[1L]][[1L]][[1L]]
  sorted_pts <- sort(unlist(pts))
  expect_length(sorted_pts, 2L)
  expect_lt(abs(sorted_pts[1L] - 0.0), 1e-9)
  expect_lt(abs(sorted_pts[2L] - 1.0), 1e-9)
})

# ---------------------------------------------------------------------------
# 1D overlay tests
# ---------------------------------------------------------------------------

test_that("1D centroid and marginal_median overlays are correct", {
  skip_without_lib()
  trace <- .trace_1d()
  w <- animate_competition_canvas(
    trace,
    voters   = .voters_1d,
    overlays = list(centroid = TRUE, marginal_median = TRUE),
    width    = 900L,
    height   = 400L
  )
  ovs <- w$x$overlays_static
  expect_true("centroid" %in% names(ovs))
  expect_true("marginal_median" %in% names(ovs))
  expect_lt(abs(ovs$centroid$x - mean(.voters_1d)),   1e-9)
  expect_lt(abs(ovs$marginal_median$x - median(.voters_1d)), 1e-9)
})

test_that("unsupported 1D overlay emits a warning and is absent", {
  skip_without_lib()
  trace <- .trace_1d()
  expect_warning(
    w <- animate_competition_canvas(
      trace,
      voters   = .voters_1d,
      overlays = list(pareto_set = TRUE),
      width    = 900L,
      height   = 400L
    ),
    regexp = "pareto_set"
  )
  ovs <- w$x[["overlays_static"]]
  if (!is.null(ovs)) expect_false("pareto_set" %in% names(ovs))
})

# ---------------------------------------------------------------------------
# WinSet 1D tests
# ---------------------------------------------------------------------------

test_that("1D winset_intervals_1d absent when compute_winset = FALSE", {
  skip_without_lib()
  trace <- .trace_1d()
  w <- animate_competition_canvas(trace, voters = .voters_1d,
                                  width = 900L, height = 400L,
                                  compute_winset = FALSE)
  expect_null(w$x[["winset_intervals_1d"]])
})

test_that("1D winset_intervals_1d present and correct shape when compute_winset = TRUE", {
  skip_without_lib()
  trace <- .trace_1d()
  w <- animate_competition_canvas(trace, voters = .voters_1d,
                                  width = 900L, height = 400L,
                                  compute_winset = TRUE)
  ws <- w$x[["winset_intervals_1d"]]
  expect_false(is.null(ws))
  # 3 frames (Round 1, Round 2, Final)
  expect_length(ws, 3L)
  # Each frame is a list of seat-holder entries
  for (frame in ws) {
    for (entry in frame) {
      expect_true(all(c("lo", "hi", "competitor_idx") %in% names(entry)))
    }
  }
})

test_that("1D winset lo <= hi when non-empty", {
  skip_without_lib()
  trace <- .trace_1d()
  w <- animate_competition_canvas(trace, voters = .voters_1d,
                                  width = 900L, height = 400L,
                                  compute_winset = TRUE)
  ws <- w$x[["winset_intervals_1d"]]
  for (frame in ws) {
    for (entry in frame) {
      if (!is.na(entry$lo) && !is.na(entry$hi)) {
        expect_lte(entry$lo, entry$hi)
      }
    }
  }
})

# ---------------------------------------------------------------------------
# Error / guard tests
# ---------------------------------------------------------------------------
# Note: the C engine only supports 1D and 2D, so a 3D guard test cannot be
# written without a mock — omitted intentionally.

File: python/tests/test_competition_canvas_1d.py
# test_competition_canvas_1d.py — Unit tests for 1D animate_competition_canvas.

import json
import re
import warnings

import numpy as np

import socialchoicelab as scl
import socialchoicelab.plots as sclp

# ---------------------------------------------------------------------------
# Fixture helpers
# ---------------------------------------------------------------------------

# 3 1D voters, 2 sticker competitors, 1 seat, 2 rounds.
# Competitor 0 at x=0 wins every round (it is the closest to 2 of 3 voters).
# n_frames = 3 (Round 1, Round 2, Final); 1 seat holder per frame.
#
# competition_run requires voters as a (n, n_dims) array for Python (it cannot
# infer n_dims from a flat 1D array without an explicit parameter).

_VOTERS_1D   = np.array([[-0.5], [0.1], [1.8]])  # shape (3, 1)
_COMPS_1D    = np.array([[ 0.0], [ 2.0]])         # shape (2, 1)
_ENGINE_1D   = scl.CompetitionEngineConfig(
    seat_count=1,
    seat_rule="plurality_top_k",
    max_rounds=2,
    step_config=scl.CompetitionStepConfig(kind="fixed", fixed_step_size=0.5),
)
_DIST_1D     = scl.DistanceConfig(salience_weights=[1.0])


def _trace_1d():
    return scl.competition_run(
        _COMPS_1D,
        ["sticker", "sticker"],
        _VOTERS_1D,
        dist_config=_DIST_1D,
        engine_config=_ENGINE_1D,
    )


def _extract_payload(html: str) -> dict:
    match = re.search(
        r"HTMLWidgets\._init\(.*?\d+,\s*\d+,\s*(\{.*\})\s*\);",
        html,
        re.DOTALL,
    )
    assert match is not None, "Could not find JSON payload in HTML."
    return json.loads(match.group(1))


# ---------------------------------------------------------------------------
# Payload shape tests
# ---------------------------------------------------------------------------


def test_1d_payload_has_n_dims_key():
    with _trace_1d() as trace:
        html = sclp.animate_competition_canvas(trace, voters=_VOTERS_1D)
    payload = _extract_payload(html)
    assert payload["n_dims"] == 1


def test_1d_payload_has_voters_x_not_voters_y():
    with _trace_1d() as trace:
        html = sclp.animate_competition_canvas(trace, voters=_VOTERS_1D)
    payload = _extract_payload(html)
    assert "voters_x" in payload
    assert "voters_y" not in payload
    assert len(payload["voters_x"]) == 3


def test_1d_payload_positions_are_length1_lists():
    with _trace_1d() as trace:
        html = sclp.animate_competition_canvas(trace, voters=_VOTERS_1D)
    payload = _extract_payload(html)
    n_frames = len(payload["positions"])
    n_comp   = len(payload["positions"][0])
    for f in range(n_frames):
        for c in range(n_comp):
            inner = payload["positions"][f][c]
            assert isinstance(inner, list), f"positions[{f}][{c}] is not a list"
            assert len(inner) == 1, f"positions[{f}][{c}] has {len(inner)} elements, expected 1"


def test_1d_payload_no_ylim():
    with _trace_1d() as trace:
        html = sclp.animate_competition_canvas(trace, voters=_VOTERS_1D)
    payload = _extract_payload(html)
    assert "ylim" not in payload


def test_1d_payload_no_winsets():
    with _trace_1d() as trace:
        html = sclp.animate_competition_canvas(
            trace, voters=_VOTERS_1D, compute_winset=False
        )
    payload = _extract_payload(html)
    assert "winsets" not in payload


def test_1d_payload_dim_names_length1():
    with _trace_1d() as trace:
        html = sclp.animate_competition_canvas(
            trace, voters=_VOTERS_1D, dim_names=("Policy",)
        )
    payload = _extract_payload(html)
    assert payload["dim_names"] == ["Policy"]


def test_1d_payload_xlim_covers_all_positions():
    with _trace_1d() as trace:
        html = sclp.animate_competition_canvas(trace, voters=_VOTERS_1D)
    payload = _extract_payload(html)
    xlim = payload["xlim"]
    assert xlim[0] <= min(_VOTERS_1D.min(), _COMPS_1D.min())
    assert xlim[1] >= max(_VOTERS_1D.max(), _COMPS_1D.max())


# ---------------------------------------------------------------------------
# IC tests
# ---------------------------------------------------------------------------


def test_1d_no_ic_by_default():
    with _trace_1d() as trace:
        html = sclp.animate_competition_canvas(trace, voters=_VOTERS_1D)
    payload = _extract_payload(html)
    assert "indifference_curves" not in payload
    assert "ic_competitor_indices" not in payload


def test_1d_ic_shape():
    with _trace_1d() as trace:
        html = sclp.animate_competition_canvas(
            trace, voters=_VOTERS_1D, compute_ic=True
        )
    payload = _extract_payload(html)
    assert "indifference_curves" in payload
    ic      = payload["indifference_curves"]
    n_frames = len(payload["positions"])   # 3
    n_voters = len(payload["voters_x"])    # 3
    assert len(ic) == n_frames
    for frame in ic:
        # one seat holder per frame (competitor 0 always wins)
        assert len(frame) == 1
        voter_curves = frame[0]
        assert len(voter_curves) == n_voters
        for pts in voter_curves:
            # each IC is 0, 1, or 2 floats
            assert isinstance(pts, list)
            assert len(pts) <= 2


def test_1d_ic_boundaries_are_symmetric():
    """For linear loss + unit weight, level set of voter at x=0.5 at utility -0.5
    should span [0.0, 1.0] — symmetric around the ideal point."""
    voters = np.array([[0.5]])   # shape (1, 1) — voter at 0.5
    comp   = np.array([[0.0]])   # shape (1, 1) — seat at 0.0

    trace = scl.competition_run(
        comp,
        ["sticker"],
        voters,
        dist_config=scl.DistanceConfig(salience_weights=[1.0]),
        engine_config=scl.CompetitionEngineConfig(
            seat_count=1,
            seat_rule="plurality_top_k",
            max_rounds=1,
            step_config=scl.CompetitionStepConfig(kind="fixed", fixed_step_size=0.0),
        ),
    )
    with trace:
        html = sclp.animate_competition_canvas(
            trace,
            voters=voters,
            compute_ic=True,
            ic_loss_config=scl.LossConfig(loss_type="linear", sensitivity=1.0),
            ic_dist_config=scl.DistanceConfig(salience_weights=[1.0]),
        )
    payload = _extract_payload(html)
    # First frame, first (only) seat holder, first (only) voter
    pts = payload["indifference_curves"][0][0][0]
    assert len(pts) == 2, f"Expected 2 boundary points, got {pts}"
    a, b = sorted(pts)
    assert abs(a - 0.0) < 1e-9, f"Left boundary {a} != 0.0"
    assert abs(b - 1.0) < 1e-9, f"Right boundary {b} != 1.0"


# ---------------------------------------------------------------------------
# 1D overlay tests
# ---------------------------------------------------------------------------


def test_1d_overlays_centroid_and_median():
    with _trace_1d() as trace:
        html = sclp.animate_competition_canvas(
            trace,
            voters=_VOTERS_1D,
            overlays={"centroid": True, "marginal_median": True},
        )
    payload = _extract_payload(html)
    ovs = payload["overlays_static"]
    assert "centroid" in ovs
    assert "marginal_median" in ovs
    expected_mean   = float(np.mean(_VOTERS_1D))
    expected_median = float(np.median(_VOTERS_1D))
    assert abs(ovs["centroid"]["x"] - expected_mean) < 1e-9
    assert abs(ovs["marginal_median"]["x"] - expected_median) < 1e-9


def test_1d_unsupported_overlay_warns_and_is_absent():
    with _trace_1d() as trace:
        with warnings.catch_warnings(record=True) as w:
            warnings.simplefilter("always")
            html = sclp.animate_competition_canvas(
                trace,
                voters=_VOTERS_1D,
                overlays={"pareto_set": True},
            )
    assert any("pareto_set" in str(warning.message) for warning in w)
    payload = _extract_payload(html)
    ovs = payload.get("overlays_static", {})
    assert "pareto_set" not in ovs


# ---------------------------------------------------------------------------
# WinSet 1D tests
# ---------------------------------------------------------------------------


def test_1d_winset_intervals_absent_when_disabled(tmp_path):
    with _trace_1d() as trace:
        html = sclp.animate_competition_canvas(
            trace, voters=_VOTERS_1D, width=900, height=400,
            path=str(tmp_path / "out.html"), compute_winset=False,
        )
    payload = _extract_payload(html)
    assert "winset_intervals_1d" not in payload


def test_1d_winset_intervals_present_and_correct_shape(tmp_path):
    with _trace_1d() as trace:
        html = sclp.animate_competition_canvas(
            trace, voters=_VOTERS_1D, width=900, height=400,
            path=str(tmp_path / "out.html"), compute_winset=True,
        )
    payload = _extract_payload(html)
    ws = payload.get("winset_intervals_1d")
    assert ws is not None, "winset_intervals_1d key missing"
    # 3 frames: Round 1, Round 2, Final
    assert len(ws) == 3
    for frame in ws:
        for entry in frame:
            assert "lo" in entry and "hi" in entry and "competitor_idx" in entry


def test_1d_winset_lo_le_hi_when_non_empty(tmp_path):
    with _trace_1d() as trace:
        html = sclp.animate_competition_canvas(
            trace, voters=_VOTERS_1D, width=900, height=400,
            path=str(tmp_path / "out.html"), compute_winset=True,
        )
    payload = _extract_payload(html)
    ws = payload["winset_intervals_1d"]
    import math
    for frame in ws:
        for entry in frame:
            lo, hi = entry["lo"], entry["hi"]
            if not math.isnan(lo) and not math.isnan(hi):
                assert lo <= hi, f"lo={lo} > hi={hi}"


# ---------------------------------------------------------------------------
# Error / guard tests
# ---------------------------------------------------------------------------
# Note: the C engine only supports 1D and 2D, so a 3D guard test cannot be
# written without a mock — omitted intentionally.
